src/eureka/S3_data_reduction/wfc3.py

```diff
@@ -33,63 +35,211 @@ def preparation_step(meta, log):
+        meta.iref = meta.iref[::-1]
+
+    # Save the reference frame for each scan direction
+    for i in meta.iref:
+        log.writelog(f"Preparing reference frame {i}...")
+        data = xrio.makeDataset()
+        data, meta, log = read(meta.segment_list[i], data, meta, log)
+        meta.n_int, meta.ny, meta.nx = data.flux.shape
+        data, meta = util.trim(data, meta)
+        # Need to add guess after trimming and before cut_aperture
+        meta.guess.append(data.guess)
+        data, meta = b2f.convert_to_e(data, meta, log)
+        meta.src_ypos = source_pos.source_pos(
+            data, meta, i, header=('SRCYPOS' in data.attrs['shdr']))
+        data['mask'] = (['time', 'y', 'x'],
+                        np.ones(data.flux.shape, dtype=bool))
+        data['mask'] = util.check_nans(data['flux'], data['mask'],
+                                       log, name='FLUX')
+        data['mask'] = util.check_nans(data['err'], data['mask'],
+                                       log, name='ERR')
+        data['mask'] = util.check_nans(data['v0'], data['mask'],
+                                       log, name='V0')
+        if hasattr(meta, 'manmask'):
+            util.manmask(data, meta, log)
+        data = flag_bg(data, meta, log)
+        data = background.BGsubtraction(data, meta, log, meta.isplots_S3)
+        cut_aperture(data, meta, log)
+        
+        # Save the reference values
+        meta.subdata_ref.append(data.flux)
+        meta.subdiffmask_ref.append(data.flatmask)
+
+    # Restore input values
+    meta.verbose = verbose
+    meta.ncpu = ncpu
+
+    return meta, log
+
+
+def conclusion_step(data, meta, log):
+    """Convert lists into arrays for saving and applies meta.sum_reads
+    if requested.
+
+    Parameters
+    ----------
+    data : Xarray Dataset
+        The Dataset object.
+    meta : eureka.lib.readECF.MetaClass
+        The current metadata object.
     log : logedit.Logedit
         The current log.
 
     Returns
     -------
+    data : Xarray Dataset
+        The updated Dataset object.
     meta : eureka.lib.readECF.MetaClass
         The updated metadata object.
     log : logedit.Logedit
         The updated log.
     """
     meta.centroids = np.array(meta.centroids)
-    meta.subflat = np.array(meta.subflat)
-    meta.flatmask = np.array(meta.flatmask)
+    meta.guess = np.array(meta.guess)
     meta.scanHeight = np.array(meta.scanHeight)
-    meta.diffmask = np.array(meta.diffmask)
-    meta.subdiffmask = np.array(meta.subdiffmask)
-    meta.drift2D = np.array(meta.drift2D)
-    meta.drift2D_int = np.array(meta.drift2D_int)
     meta.subdata_ref = np.array(meta.subdata_ref)
-    meta.diffmask_ref = np.array(meta.diffmask_ref)
+    meta.subdiffmask_ref = np.array(meta.subdiffmask_ref)
 
-    return meta, log
+    if meta.sum_reads:
+        # Sum each read from a scan together
+        data, meta = sum_reads(data, meta)
+
+    return data, meta, log
+
+
+def sum_reads(data, meta):
```

Comment:
Ah okay, I did it there in order to compare to the plots you'd shared with me, but it does make more sense to do that in Stage 4 for several reasons.

Reply:
Alright, I moved this step to Stage 4. And to make this a bit easier on me, I changed S4 to actually use optmask instead of using numpy masked arrays as had been on our to-do list for a while. The end result should still be the same, but with better tracking of the mask between stages.